packages/swing-store/src/importer.js

```diff
@@ -47,158 +40,109 @@ export async function importSwingStore(exporter, dirPath = null, options = {}) {
-          const metadata = JSON.parse(value);
-          vatInfo.transcriptKey = artifactKey(tag, vatID, metadata.startPos);
-          artifactMetadata.set(vatInfo.transcriptKey, metadata);
-        }
-        vatArtifacts.set(vatID, vatInfo);
-      } else {
-        artifactMetadata.set(artifactKey(tag, vatID, pos), JSON.parse(value));
-      }
+    } else if (value == null) {
+      allMetadata.delete(key);
     } else {
-      Fail`unknown artifact type tag ${q(tag)} on import`;
+      allMetadata.set(key, value);
     }
   }
 
-  // At this point we should have acquired the entire KV store state, plus
-  // sufficient metadata to identify the complete set of artifacts we'll need to
-  // fetch along with the information required to validate each of them after
-  // fetching.
-  //
-  // Depending on how the export was parameterized, the metadata may also include
-  // information about historical artifacts that we might or might not actually
-  // fetch depending on how this import was parameterized
-
-  // Fetch the set of current artifacts.
-
-  // Keep track of fetched artifacts in this set so we don't fetch them a second
-  // time if we are trying for historical artifacts also.
-  const fetchedArtifacts = new Set();
-
-  for await (const [vatID, vatInfo] of vatArtifacts.entries()) {
-    // For each vat, we *must* have a transcript span.  If this is not the very
-    // first transcript span in the history of that vat, then we also must have
-    // a snapshot for the state of the vat immediately prior to when the
-    // transcript span begins.
-    vatInfo.transcriptKey ||
-      Fail`missing current transcript key for vat ${q(vatID)}`;
-    const transcriptInfo = artifactMetadata.get(vatInfo.transcriptKey);
-    transcriptInfo || Fail`missing transcript metadata for vat ${q(vatID)}`;
-    let snapshotInfo;
-    if (vatInfo.snapshotKey) {
-      snapshotInfo = artifactMetadata.get(vatInfo.snapshotKey);
-      snapshotInfo || Fail`missing snapshot metadata for vat ${q(vatID)}`;
-    }
-    if (!snapshotInfo) {
-      transcriptInfo.startPos === 0 ||
-        Fail`missing current snapshot for vat ${q(vatID)}`;
+  // Now take each metadata record and install the stub/pruned entry
+  // into the DB.
+
+  for (const [key, value] of allMetadata.entries()) {
+    const [tag, ...pieces] = key.split('.');
+    if (tag === 'bundle') {
+      const bundleID = pieces[0];
+      internal.bundleStore.addBundleRecord(bundleID);
+    } else if (tag === 'snapshot') {
+      const [_vatID, endPos] = pieces;
+      if (endPos !== 'current') {
+        internal.snapStore.addSnapshotRecord(JSON.parse(value));
+      }
+      // metadata['snapshot.v1.current'] = 'snapshot.v1.5' , i.e. it
+      // points to the name of the current artifact. We could
+      // conceivably remember this and compare it against the .inUse
+      // property of that record, but it's not worth the effort (we
+      // might encounter the records in either order).
```

Comment:
Without it, the only way to locate the active snapshot for each vat from the export data is to parse the value of every `snapshot.v1.*` entry, looking for the one with `inUse: 1`, which is O(N) in the number of snapshots.

With it, we can do it in two steps: read the value of `snapshot.v1.current`, then use that value as the second key to look up.

We aren't doing this sort of query against the export data when we do an import, but I wanted to keep the format easy to parse in case some other need came along.

I suppose a valid question is what happens if `snapshot.vNN.current` is missing, or the value is malformed, or the value points to a snapshot record which has `inUse: 0`. And I suppose the answer is that I don't need the importer to also be a thorough correctness checker :). This one record isn't necessary to do an import, so all I really want to do is ignore it.